numpy/ma/core.py

```diff
@@ -3305,7 +3305,8 @@ def _scalar_heuristic(arr, elem):
                     # which should have shape () to avoid breaking several
                     # methods. There is no great way out, so set to
                     # first element.  See issue #6723.
-                    if dout._fill_value.ndim > 0:
+                    dtype = self._fill_value.dtype[indx]
+                    if dtype.ndim > 0:
```

Comment:
To be clear, I'd suggest we change these lines to
```suggestion
                    # something like gh-15895 has happened if this fails
                    assert isinstance(dout._fill_value, np.ndarray)
                    if dout._fill_value.ndim > 0:
```